File: switch/switch1.ts
type A = {
    readonly a: string;
};

type B = {
    readonly b: string;
};

function value(useA: boolean): A | B {
    if (useA) {
        return { a: "A" };
    } else {
        return { b: "B" };
    }
}

function key(useA: boolean): keyof A | keyof B {
    if (useA) {
        return 'a';
    } else {
        return 'b';
    }
}

function test(useA: boolean): string {
    return value(useA)[key(useA)];
}

export { };

File: switch/switch1.hs
data A = A
data B = B

value :: Bool -> Either A B
value True = Left A
value False = Right B

func :: Bool -> Either (A -> String) (B -> String)
func True = Left $ \A -> "A"
func False = Right $ \B -> "B"

test :: Bool -> String
test useA = case (value useA, func useA) of
  (Left a, Left fa) -> fa a
  (Right b, Right fb) -> fb b
  _ -> error "Never happens"
